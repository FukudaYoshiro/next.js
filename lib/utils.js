export function warn (message) {
  if (process.env.NODE_ENV !== 'production') {
    console.error(message)
  }
}

export function execOnce (fn) {
  let used = false
  return (...args) => {
    if (!used) {
      used = true
      fn.apply(this, args)
    }
  }
}

export function deprecated (fn, message) {
  // else is used here so that webpack/uglify will remove the code block depending on the build environment
  if (process.env.NODE_ENV === 'production') {
    return fn
  } else {
    let warned = false
    const newFn = function (...args) {
      if (!warned) {
        warned = true
        console.error(message)
      }
      return fn.apply(this, args)
    }

    // copy all properties
    Object.assign(newFn, fn)

    return newFn
  }
}

export function printAndExit (message, code = 1) {
  if (code === 0) {
    console.log(message)
  } else {
    console.error(message)
  }

  process.exit(code)
}

export function getDisplayName (Component) {
  if (typeof Component === 'string') {
    return Component
  }

  return Component.displayName || Component.name || 'Unknown'
}

export function isResSent (res) {
  return res.finished || res.headersSent
}

export async function loadGetInitialProps (Component, ctx) {
  if (process.env.NODE_ENV !== 'production') {
    if (Component.prototype && Component.prototype.getInitialProps) {
      const compName = getDisplayName(Component)
      const message = `"${compName}.getInitialProps()" is defined as an instance method - visit https://err.sh/zeit/next.js/get-initial-props-as-an-instance-method for more information.`
      throw new Error(message)
    }
  }

  if (!Component.getInitialProps) return {}

  const props = await Component.getInitialProps(ctx)

  if (ctx.res && isResSent(ctx.res)) {
    return props
  }

  if (!props) {
    const compName = getDisplayName(Component)
    const message = `"${compName}.getInitialProps()" should resolve to an object. But found "${props}" instead.`
    throw new Error(message)
  }

  return props
}

export function getLocationOrigin () {
  const { protocol, hostname, port } = window.location
  return `${protocol}//${hostname}${port ? ':' + port : ''}`
}

export function getURL () {
  const { href } = window.location
  const origin = getLocationOrigin()
  return href.substring(origin.length)
}
