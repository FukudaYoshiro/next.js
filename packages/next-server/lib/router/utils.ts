export function getRouteRegex(
  route: string,
): { re: RegExp; groups: { [groupName: string]: number } } {
  const escapedRoute = (route.replace(/\/$/, '') || '/').replace(
    /[|\\{}()[\]^$+*?.-]/g,
    '\\$&',
  )

  const groups: { [groupName: string]: number } = {}
  let groupIndex = 1

  const parameterizedRoute = escapedRoute.replace(
    /\/\\\$([^\/]+?)(?=\/|$)/g,
    (_, $1) => (
      (groups[
        $1.replace(/\\\$$/, '').replace(/\\([|\\{}()[\]^$+*?.-])/g, '$1')
      ] = groupIndex++),
      $1.lastIndexOf('$') === $1.length - 1 ? '(?:/([^/]+?))?' : '/([^/]+?)'
    ),
  )

  return {
    re: new RegExp('^' + parameterizedRoute + '(?:/)?$', 'i'),
    groups,
  }
}

export function getRouteMatch(route: string) {
  const { re, groups } = getRouteRegex(route)
  return (pathname: string | undefined) => {
    const routeMatch = re.exec(pathname!)
    if (!routeMatch) {
      return false
    }

    const params: { [paramName: string]: string } = {}

    Object.keys(groups).forEach((slugName: string) => {
      const m = routeMatch[groups[slugName]]
      if (m !== undefined) {
        params[slugName] = decodeURIComponent(m)
      }
    })
    return params
  }
}
