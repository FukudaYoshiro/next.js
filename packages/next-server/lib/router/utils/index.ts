export { getRouteMatcher } from './route-matcher'
export { getRouteRegex } from './route-regex'
export { getSortedRoutes } from './sorted-routes'
