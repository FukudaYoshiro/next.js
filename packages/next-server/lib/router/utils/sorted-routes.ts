class UrlNode {
  placeholder: boolean = true
  children: Map<string, UrlNode> = new Map()
  slugName: string | null = null

  get hasSlug() {
    return this.slugName != null
  }

  insert(urlPath: string): void {
    this._insert(urlPath.split('/').filter(Boolean))
  }

  smoosh(): string[] {
    return this._smoosh()
  }

  private _smoosh(prefix: string = '/'): string[] {
    const childrenPaths = [...this.children.keys()].sort()
    if (this.hasSlug) {
      childrenPaths.splice(childrenPaths.indexOf('$'), 1)
    }

    const routes = childrenPaths
      .map(c => this.children.get(c)!._smoosh(`${prefix}${c}/`))
      .reduce((prev, curr) => [...prev, ...curr], [])

    if (this.hasSlug) {
      routes.push(
        ...this.children.get('$')!._smoosh(`${prefix}$${this.slugName}/`)
      )
    }

    if (!this.placeholder) {
      routes.unshift(prefix === '/' ? '/' : prefix.slice(0, -1))
    }

    return routes
  }

  private _insert(urlPaths: string[]): void {
    if (urlPaths.length === 0) {
      this.placeholder = false
      return
    }

    let [nextSegment] = urlPaths
    if (nextSegment.startsWith('$')) {
      const slugName = nextSegment.substring(1)
      if (this.hasSlug && slugName !== this.slugName) {
        throw new Error(
          'You cannot use different slug names for the same dynamic path.'
        )
      }

      this.slugName = slugName
      nextSegment = '$'
    }

    if (!this.children.has(nextSegment)) {
      this.children.set(nextSegment, new UrlNode())
    }

    this.children.get(nextSegment)!._insert(urlPaths.slice(1))
  }
}

export function getSortedRoutes(normalizedPages: string[]): string[] {
  const root = new UrlNode()
  normalizedPages.forEach(page => root.insert(page))
  return root.smoosh()
}
