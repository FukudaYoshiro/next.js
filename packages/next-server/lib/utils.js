import { format } from 'url'

export function execOnce (fn) {
  let used = false
  return (...args) => {
    if (!used) {
      used = true
      fn.apply(this, args)
    }
  }
}

export function getLocationOrigin () {
  const { protocol, hostname, port } = window.location
  return `${protocol}//${hostname}${port ? ':' + port : ''}`
}

export function getURL () {
  const { href } = window.location
  const origin = getLocationOrigin()
  return href.substring(origin.length)
}

export function getDisplayName (Component) {
  if (typeof Component === 'string') {
    return Component
  }

  return Component.displayName || Component.name || 'Unknown'
}

export function isResSent (res) {
  return res.finished || res.headersSent
}

export async function loadGetInitialProps (Component, ctx) {
  if (process.env.NODE_ENV !== 'production') {
    if (Component.prototype && Component.prototype.getInitialProps) {
      const message = `"${getDisplayName(Component)}.getInitialProps()" is defined as an instance method - visit https://err.sh/zeit/next.js/get-initial-props-as-an-instance-method for more information.`
      throw new Error(message)
    }
  }

  if (!Component.getInitialProps) return {}

  const props = await Component.getInitialProps(ctx)

  if (ctx.res && isResSent(ctx.res)) {
    return props
  }

  if (!props) {
    const message = `"${getDisplayName(Component)}.getInitialProps()" should resolve to an object. But found "${props}" instead.`
    throw new Error(message)
  }

  return props
}

export const urlObjectKeys = ['auth', 'hash', 'host', 'hostname', 'href', 'path', 'pathname', 'port', 'protocol', 'query', 'search', 'slashes']

export function formatWithValidation (url, options) {
  if (process.env.NODE_ENV === 'development') {
    if (url !== null && typeof url === 'object') {
      Object.keys(url).forEach((key) => {
        if (urlObjectKeys.indexOf(key) === -1) {
          console.warn(`Unknown key passed via urlObject into url.format: ${key}`)
        }
      })
    }
  }

  return format(url, options)
}
