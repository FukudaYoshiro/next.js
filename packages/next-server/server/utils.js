import { BLOCKED_PAGES } from 'next-server/constants'

const internalPrefixes = [
  /^\/_next\//,
  /^\/static\//
]

export function isInternalUrl (url) {
  for (const prefix of internalPrefixes) {
    if (prefix.test(url)) {
      return true
    }
  }

  return false
}

export function isBlockedPage (pathname) {
  return (BLOCKED_PAGES.indexOf(pathname) !== -1)
}
