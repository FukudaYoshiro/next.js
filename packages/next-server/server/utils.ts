import { BLOCKED_PAGES } from 'next-server/constants'

const internalPrefixes = [
  /^\/_next\//,
  /^\/static\//,
]

export function isInternalUrl(url: string): boolean {
  for (const prefix of internalPrefixes) {
    if (prefix.test(url)) {
      return true
    }
  }

  return false
}

export function isBlockedPage(pathname: string): boolean {
  return (BLOCKED_PAGES.indexOf(pathname) !== -1)
}
