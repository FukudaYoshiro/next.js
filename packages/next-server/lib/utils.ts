import { format, UrlObject, URLFormatOptions } from 'url'
import { ServerResponse, IncomingMessage } from 'http';
import { ComponentType } from 'react'
import { ParsedUrlQuery } from 'querystring'
import { ManifestItem } from '../server/get-dynamic-import-bundles'
import { BaseRouter } from './router/router'

/**
 * Types used by both next and next-server
 */
export type NextComponentType<C extends BaseContext = any, P = any, CP = {}> = ComponentType<CP> & {
  getInitialProps?(context: C): Promise<P>,
}

export type DocumentType = NextComponentType<DocumentContext, DocumentInitialProps, DocumentProps>

export type AppType = NextComponentType<AppContextType, AppInitialProps, AppPropsType>

export type Enhancer<C> = (Component: C) => C

export type ComponentsEnhancer =
  | { enhanceApp?: Enhancer<AppType>; enhanceComponent?: Enhancer<NextComponentType> }
  | Enhancer<NextComponentType>

export type RenderPageResult = { html: string, head?: Array<JSX.Element | null>, dataOnly?: true }

export type RenderPage = (options?: ComponentsEnhancer) => RenderPageResult | Promise<RenderPageResult>

export type BaseContext = {
  res?: ServerResponse
  [k: string]: any,
}

export type INEXTDATA = {
  dataManager: string
  props: any
  page: string
  query: ParsedUrlQuery
  buildId: string
  dynamicBuildId: boolean
  assetPrefix?: string
  runtimeConfig?: { [key: string]: any }
  nextExport?: boolean
  dynamicIds?: string[]
  err?: Error & { statusCode?: number },
}

export interface IContext {
  err?: Error & { statusCode?: number } | null
  req?: IncomingMessage
  res?: ServerResponse
  pathname: string
  query: ParsedUrlQuery
  asPath?: string
}

export type AppContextType<R extends BaseRouter = BaseRouter> = {
  Component: NextComponentType<IContext>
  router: R
  ctx: IContext,
}

export type AppInitialProps = {
  pageProps: any,
}

export type AppPropsType<R extends BaseRouter = BaseRouter> = AppInitialProps & {
  Component: NextComponentType<IContext>
  router: R,
}

export type DocumentContext = IContext & {
  renderPage: RenderPage,
}

export type DocumentInitialProps = RenderPageResult & {
  styles?: React.ReactElement[],
}

export type DocumentProps = DocumentInitialProps & {
  __NEXT_DATA__: INEXTDATA
  dangerousAsPath: string
  ampPath: string
  amphtml: boolean
  hasAmp: boolean
  staticMarkup: boolean
  devFiles: string[]
  files: string[]
  dynamicImports: ManifestItem[]
  assetPrefix?: string,
}

/**
 * Utils
 */

export function execOnce(this: any, fn: () => any) {
  let used = false
  return (...args: any) => {
    if (!used) {
      used = true
      fn.apply(this, args)
    }
  }
}

export function getLocationOrigin() {
  const { protocol, hostname, port } = window.location
  return `${protocol}//${hostname}${port ? ':' + port : ''}`
}

export function getURL() {
  const { href } = window.location
  const origin = getLocationOrigin()
  return href.substring(origin.length)
}

export function getDisplayName(Component: ComponentType<any>) {
  return typeof Component === 'string' ? Component : (Component.displayName || Component.name || 'Unknown')
}

export function isResSent(res: ServerResponse) {
  return res.finished || res.headersSent
}

export async function loadGetInitialProps<C extends BaseContext, P = any, CP = {}>(Component: NextComponentType<C, P, CP>, ctx: C): Promise<P | null> {
  if (process.env.NODE_ENV !== 'production') {
    if (Component.prototype && Component.prototype.getInitialProps) {
      const message = `"${getDisplayName(Component)}.getInitialProps()" is defined as an instance method - visit https://err.sh/zeit/next.js/get-initial-props-as-an-instance-method for more information.`
      throw new Error(message)
    }
  }

  if (!Component.getInitialProps) {
    if (ctx.res && ctx.res.setHeader) {
      ctx.res.setHeader('Cache-Control', 'stale-while-revalidate=30')
    }
    return null
  }

  const props = await Component.getInitialProps(ctx)

  if (ctx.res && isResSent(ctx.res)) {
    return props
  }

  if (!props) {
    const message = `"${getDisplayName(Component)}.getInitialProps()" should resolve to an object. But found "${props}" instead.`
    throw new Error(message)
  }

  return props
}

export const urlObjectKeys = ['auth', 'hash', 'host', 'hostname', 'href', 'path', 'pathname', 'port', 'protocol', 'query', 'search', 'slashes']

export function formatWithValidation(url: UrlObject, options?: URLFormatOptions) {
  if (process.env.NODE_ENV === 'development') {
    if (url !== null && typeof url === 'object') {
      Object.keys(url).forEach((key) => {
        if (urlObjectKeys.indexOf(key) === -1) {
          console.warn(`Unknown key passed via urlObject into url.format: ${key}`)
        }
      })
    }
  }

  return format(url as any, options)
}
