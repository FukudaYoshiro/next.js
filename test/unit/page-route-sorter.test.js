/* eslint-env jest */
import { getSortedRoutes } from 'next-server/dist/lib/router/utils/sorted-routes'

describe('getSortedRoutes', () => {
  it('does not add extra routes', () => {
    expect(getSortedRoutes(['/posts'])).toEqual(['/posts'])

    expect(getSortedRoutes(['/posts/$id'])).toEqual(['/posts/$id'])
    expect(getSortedRoutes(['/posts/$id/foo'])).toEqual(['/posts/$id/foo'])

    expect(getSortedRoutes(['/posts/$id/$foo/bar'])).toEqual([
      '/posts/$id/$foo/bar'
    ])
    expect(getSortedRoutes(['/posts/$id/baz/$foo/bar'])).toEqual([
      '/posts/$id/baz/$foo/bar'
    ])
  })

  it('correctly sorts required slugs', () => {
    expect(
      getSortedRoutes([
        '/posts',
        '/$root-slug',
        '/',
        '/posts/$id',
        '/blog/$id/comments/$cid',
        '/blog/$id',
        '/foo/$d/bar/baz/$f',
        '/apples/$ab/$cd/ef'
      ])
    ).toMatchSnapshot()
  })
})
