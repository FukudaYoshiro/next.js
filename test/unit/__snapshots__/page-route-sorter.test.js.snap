// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSortedRoutes correctly sorts required slugs 1`] = `
Array [
  "/",
  "/apples/[ab]/[cd]/ef",
  "/blog/[id]",
  "/blog/[id]/comments/[cid]",
  "/foo/[d]/bar/baz/[f]",
  "/posts",
  "/posts/[id]",
  "/[root-slug]",
]
`;
